### ISSUE TITLE
[BUG] $ref resolution fails for OpenAPI spec on schemathesis 3.18.3

### ISSUE THREAD

**Issue by jmidt** [schemathesis/schemathesis#1697]
**Checklist**

- [x] I checked the [FAQ section](https://schemathesis.readthedocs.io/en/stable/faq.html#frequently-asked-questions) of the documentation
- [x] I looked for similar issues in the [issue tracker](https://github.com/schemathesis/schemathesis/issues)

**Describe the bug**
Including the OpenAPI spec itself in the spec causes Schemathesis to run forever in the resolve_all() step of the schema parsing. This would be a use case if you supply your own API spec through your API, and note the OpenAPI spec as the schema.

**To Reproduce**

- Use the API spec below.
- Read the file with `schema = schemathesis.from_path(file_path)`
- Accessing e.g. `schema["/"]` will cause schemathesis to hang forever in the resolve_all() step, consuming more and more memory. I let it run up to at least 30G of memory consumed with no sign of continuing.

```yaml
{
    openapi: 3.0.3
    info:
      version: 1.0.0
      title: My API
      description: My HTTP interface.
    
    paths:
      # Get this specification
      /:
        get:
          summary: OpenAPI description (this document)
          tags:
            - Introspection
          operationId: getApiDescription
          responses:
            "200":
              description: OK
              content:
                application/openapi+json:
                  schema:
                    $ref: "https://raw.githubusercontent.com/OAI/OpenAPI-Specification/aa91a19c43f8a12c02efa42d64794e396473f3b1/schemas/v3.0/schema.json#"
}
```

**Expected behavior**
I would expect schemathesis to be able to parse the OpenAPI spec itself.

**Environment (please complete the following information):**

- OS: Ubuntu
- Python version: 3.10
- Schemathesis version: 3.18.3
- Spec version: 3.0.3

**Additional context**
It worked on Schemathesis 3.15.6.

**Comment by Stranger6667** [comment]
Indeed, it is bad behavior - I'll investigate what's the root cause. In some cases, we fallback to inlining schemas up to some recursion level as `hypothesis-jsonschema` does not yet support recursive references. Implementing it properly is on my todo list for the next following months (as it is needed to support Open API 3.1 properly), so implementing this will solve one of the possible causes.

**Comment by Stranger6667** [comment]
Fixed in one of the recent commits that removed `resolve_all` & reworked all the reference handling.
